### ISSUE TITLE
Quantum 2.3.1 Crashing Application When Clustering

### ISSUE THREAD

**Issue by zacksiri** [quantum-elixir/quantum-core#361]
As I've mentioned here #345 , the new swarm implementation crashes my app.

```
17:11:53.806 [info] [libcluster:studio_cluster] connected to :"studio@240.4.0.28"
17:11:53.808 [info] [swarm on studio@240.5.0.106] [tracker:ensure_swarm_started_on_remote_node] nodeup studio@240.4.0.28
17:11:53.809 [info] [swarm on studio@240.5.0.106] [tracker:cluster_wait] joining cluster..
17:11:53.809 [info] [swarm on studio@240.5.0.106] [tracker:cluster_wait] syncing with studio@240.4.0.28
17:11:53.810 [info] [swarm on studio@240.5.0.106] [tracker:awaiting_sync_ack] received sync acknowledgement from studio@240.4.0.28, syncing with remote registry
17:11:53.810 [info] [swarm on studio@240.5.0.106] [tracker:awaiting_sync_ack] local synchronization with studio@240.4.0.28 complete!
17:11:53.810 [info] [swarm on studio@240.5.0.106] [tracker:resolve_pending_sync_requests] pending sync requests cleared
17:11:53.855 [info] Application studio exited: Studio.Application.start(:normal, []) returned an error: shutdown: failed to start child: Studio.Scheduler
    ** (EXIT) shutdown: failed to start child: Quantum.TaskRegistry
        ** (EXIT) already started: #PID<43527.2717.0>
{"Kernel pid terminated",application_controller,"{application_start_failure,studio,{{shutdown,{failed_to_start_child,'Elixir.Studio.Scheduler',{shutdown,{failed_to_start_child,'Elixir.Quantum.TaskRegistry',{already_started,<43527.2717.0>}}}}},{'Elixir.Studio.Application',start,[normal,[]]}}}"}
Kernel pid terminated (application_controller) ({application_start_failure,studio,{{shutdown,{failed_to_start_child,'Elixir.Studio.Scheduler',{shutdown,{failed_to_start_child,'Elixir.Quantum.TaskRegis
```

**Comment by maennchen** [comment]
Do you have multiple Schedulers in your app? If so, dies it work if you disable all except one?

**Comment by zacksiri** [comment]
How do I disable one when they start along with the app? And it’s the same app that join into a cluster. If I don’t cluster or only have 1 node it works fine.

**Comment by zacksiri** [comment]
@maennchen so I disabled `global: true` and it seems to be working fine. is that intentional / how it's supposed to be used? now that we have swarm it can ensure 2 jobs aren't started by 2 scheduler on different nodes?

**Comment by maennchen** [comment]
I‘ll try to figure out the problem later today. Is your project OS so that I could try it myself?

**Comment by zacksiri** [comment]
@maennchen there is just 1 scheduler, in the app, but yeah it runs 2 jobs as expected. it's not OS but I can add you as a member do you have a gitlab account?

**Comment by maennchen** [comment]
@zacksiri I could not reproduce the issue in any of my projects. Therefore I'll need to see the code to investigate further.

I've just registered on GitLab with the same username as here: `@maennchen`.

**Comment by zacksiri** [comment]
@maennchen I've added you on gitlab

**Comment by maennchen** [comment]
@zacksiri In some of the I use it. I don't think that this will create a problem.

**Comment by zacksiri** [comment]
@maennchen ok sure thing the upstream should not effect the testing. btw how are you testing? let me replicate your case.

**Comment by maennchen** [comment]
Can you try with #364?

**Comment by zacksiri** [comment]
Let me try..

**Comment by maennchen** [comment]
@zacksiri :(

Then I'll have to look at it tomorrow... I don't have time currently today to go deeper into the rabbit hole. I'll do some further testing and keep you up to date.

**Comment by zacksiri** [comment]
@maennchen no worries, thx for the effort, I'll revert to 2.2.7 for now.

**Comment by zacksiri** [comment]
@maennchen I may have messed something up in my deploy, let me do more testing with an update before you spend more time on it.

**Comment by redink** [comment]
reproduce step: https://github.com/redink/t_quantum/tree/master/test

**Comment by zacksiri** [comment]
@maennchen After more thorough testing, i can confirm that #364 doesn't resolve the issue yet.

**Comment by maennchen** [comment]
@redink Yes, that's what I fixed in the PR: https://github.com/quantum-elixir/quantum-core/pull/364/files#diff-97efc738bed75d9e25c723356d9cd32c

**Comment by redink** [comment]
I updated the deps in my demo (https://github.com/redink/t_quantum/tree/master/test), but got the new error information:

```
iex(b@127.0.0.1)4> :net_adm.ping(:"a@127.0.0.1")
:pong

20:30:17.626 [info]  [swarm on b@127.0.0.1] [tracker:ensure_swarm_started_on_remote_node] nodeup a@127.0.0.1
iex(b@127.0.0.1)5> 
20:30:17.626 [info]  [swarm on b@127.0.0.1] [tracker:cluster_wait] joining cluster..
 
20:30:17.626 [info]  [swarm on b@127.0.0.1] [tracker:cluster_wait] found connected nodes: [:"a@127.0.0.1"]
 
20:30:17.626 [info]  [swarm on b@127.0.0.1] [tracker:cluster_wait] selected sync node: a@127.0.0.1
 
20:30:17.627 [info]  [swarm on b@127.0.0.1] [tracker:syncing] syncing to a@127.0.0.1 based on node precedence
 
20:30:17.628 [info]  [swarm on b@127.0.0.1] [tracker:awaiting_sync_ack] received sync acknowledgement from a@127.0.0.1, syncing with remote registry
 
20:30:17.628 [debug] [swarm on b@127.0.0.1] [tracker:sync_registry] remote and local view of Test.Scheduler.TaskRegistry conflict, but remote is correct, resolving..
 
20:30:17.629 [debug] [swarm on b@127.0.0.1] [tracker:sync_registry] remote and local view of Test.Scheduler.JobBroadcaster conflict, but remote is correct, resolving..
 
20:30:17.630 [debug] [swarm on b@127.0.0.1] [tracker:sync_registry] local tracker is missing :"Elixir.Test.Scheduler.ClusterTaskSupervisorRegistry.Group.#Reference<0.3142663777.546832387.174217>", adding to registry
 
20:30:17.630 [info]  [swarm on b@127.0.0.1] [tracker:awaiting_sync_ack] local synchronization with a@127.0.0.1 complete!

20:30:17.630 [info]  [swarm on b@127.0.0.1] [tracker:resolve_pending_sync_requests] pending sync requests cleared

20:30:17.630 [debug] [swarm on b@127.0.0.1] [tracker:handle_replica_event] replica event: untrack #PID<16592.365.0>
 
20:30:17.630 [warn]  [swarm on b@127.0.0.1] [tracker:handle_cast] unrecognized cast: {:untrack, #PID<0.188.0>}

20:30:17.630 [debug] [swarm on b@127.0.0.1] [tracker:handle_monitor] #PID<0.188.0> is down: :shutdown, but no registration found, ignoring..

20:30:17.630 [warn]  [swarm on b@127.0.0.1] [tracker:handle_cast] unrecognized cast: {:track, Test.Scheduler.TaskRegistry, #PID<16592.363.0>, %{}}
 
20:30:17.630 [warn]  [swarm on b@127.0.0.1] [tracker:handle_cast] unrecognized cast: {:untrack, #PID<0.189.0>}
 
20:30:17.630 [warn]  [swarm on b@127.0.0.1] [tracker:handle_cast] unrecognized cast: {:track, Test.Scheduler.JobBroadcaster, #PID<16592.364.0>, %{}}
 
20:30:17.630 [debug] [swarm on b@127.0.0.1] [tracker:handle_monitor] Test.Scheduler.ExecutionBroadcaster is down: :shutdown
 
20:30:17.631 [debug] [swarm on b@127.0.0.1] [tracker:handle_monitor] :"Elixir.Test.Scheduler.ClusterTaskSupervisorRegistry.Group.#Reference<0.3155748787.2694316036.156468>" is down: :shutdown
 
20:30:17.631 [debug] [swarm on b@127.0.0.1] [tracker:handle_replica_event] replica event: untrack #PID<0.190.0>
 
20:30:17.631 [debug] [swarm on b@127.0.0.1] [tracker:handle_replica_event] replica event: untrack #PID<0.186.0>
 
20:30:17.632 [debug] [swarm on b@127.0.0.1] [tracker:handle_monitor] Test.Scheduler.JobBroadcaster is down: :shutdown
 
20:30:17.632 [debug] [swarm on b@127.0.0.1] [tracker:handle_monitor] Test.Scheduler.TaskRegistry is down: :shutdown
 
20:30:17.632 [debug] [swarm on b@127.0.0.1] [tracker:handle_monitor] :"Elixir.Test.Scheduler.ClusterTaskSupervisorRegistry.Group.#Reference<0.3142663777.546832387.174217>" is down: :shutdown
 
20:30:17.633 [debug] [swarm on b@127.0.0.1] [tracker:handle_replica_event] replica event: untrack #PID<16592.364.0>
 
20:30:17.633 [debug] [swarm on b@127.0.0.1] [tracker:handle_call] registering #PID<0.210.0> as :"Elixir.Test.Scheduler.ClusterTaskSupervisorRegistry.Group.#Reference<0.3155748787.2694316036.156605>", with metadata %{}
 
20:30:17.634 [debug] [swarm on b@127.0.0.1] [tracker:handle_replica_event] replica event: untrack #PID<16592.363.0>
 
20:30:17.634 [debug] [swarm on b@127.0.0.1] [tracker:handle_replica_event] replica event: untrack #PID<16592.361.0>
 
20:30:17.634 [debug] [swarm on b@127.0.0.1] [tracker:handle_call] add_meta {Test.Scheduler.ClusterTaskSupervisorRegistry.Group, true} to #PID<0.210.0>
 
20:30:17.638 [debug] [swarm on b@127.0.0.1] [tracker:handle_replica_event] replicating registration for :"Elixir.Test.Scheduler.ClusterTaskSupervisorRegistry.Group.#Reference<0.3142663777.546832386.173074>" (#PID<16592.480.0>) locally
 
20:30:17.638 [debug] [swarm on b@127.0.0.1] [tracker:handle_replica_event] replica event: add_meta {Test.Scheduler.ClusterTaskSupervisorRegistry.Group, true} to #PID<16592.480.0>
 
20:30:17.638 [debug] [swarm on b@127.0.0.1] [tracker:handle_replica_event] replicating registration for Test.Scheduler.TaskRegistry (#PID<16592.482.0>) locally
 
20:30:17.638 [debug] [swarm on b@127.0.0.1] [tracker:handle_replica_event] replicating registration for Test.Scheduler.JobBroadcaster (#PID<16592.483.0>) locally
 
20:30:17.638 [debug] [swarm on b@127.0.0.1] [tracker:handle_replica_event] replicating registration for Test.Scheduler.ExecutionBroadcaster (#PID<16592.485.0>) locally
 
20:30:17.638 [debug] [swarm on b@127.0.0.1] [tracker:handle_monitor] :"Elixir.Test.Scheduler.ClusterTaskSupervisorRegistry.Group.#Reference<0.3155748787.2694316036.156605>" is down: :shutdown
 
20:30:17.639 [debug] [swarm on b@127.0.0.1] [tracker:handle_replica_event] replica event: untrack #PID<0.210.0>
 
20:30:17.641 [info]  GenStage consumer #PID<0.190.0> is stopping after receiving cancel from producer #PID<0.189.0> with reason: :shutdown

 
20:30:17.641 [info]  GenStage consumer Test.Scheduler.ExecutorSupervisor is stopping after receiving cancel from producer #PID<0.190.0> with reason: :shutdown


20:30:17.641 [info]  GenStage consumer Test.Scheduler.ExecutorSupervisor is stopping after receiving cancel from producer #PID<0.190.0> with reason: :noproc


20:30:17.641 [debug] [swarm on b@127.0.0.1] [tracker:handle_call] registering #PID<0.215.0> as :"Elixir.Test.Scheduler.ClusterTaskSupervisorRegistry.Group.#Reference<0.3155748787.2694316038.155214>", with metadata %{}
 
20:30:17.642 [debug] [swarm on b@127.0.0.1] [tracker:handle_call] add_meta {Test.Scheduler.ClusterTaskSupervisorRegistry.Group, true} to #PID<0.215.0>
 
20:30:17.644 [debug] [swarm on b@127.0.0.1] [tracker:handle_monitor] :"Elixir.Test.Scheduler.ClusterTaskSupervisorRegistry.Group.#Reference<0.3155748787.2694316038.155214>" is down: :shutdown
 
20:30:17.645 [debug] [swarm on b@127.0.0.1] [tracker:handle_replica_event] replica event: untrack #PID<0.215.0>
 
20:30:17.647 [error] GenServer Test.Scheduler.ExecutorSupervisor terminating
** (stop) no process: the process is not alive or there's no process currently associated with the given name, possibly because its application isn't started
Last message: {:DOWN, #Reference<0.3155748787.2694316036.156584>, :process, #PID<0.190.0>, :noproc}
State: %ConsumerSupervisor{args: %{cluster_task_supervisor_registry: Test.Scheduler.ClusterTaskSupervisorRegistry, debug_logging: true, execution_broadcaster: {:via, :swarm, Test.Scheduler.ExecutionBroadcaster}, task_registry: {:via, :swarm, Test.Scheduler.TaskRegistry}, task_supervisor: Test.Scheduler.Task.Supervisor}, children: %{}, max_restarts: 3, max_seconds: 5, mod: Quantum.ExecutorSupervisor, name: Test.Scheduler.ExecutorSupervisor, producers: %{}, restarting: 0, restarts: [], strategy: :one_for_one, template: {Quantum.Executor, {Quantum.Executor, :start_link, [%{cluster_task_supervisor_registry: Test.Scheduler.ClusterTaskSupervisorRegistry, debug_logging: true, task_registry: {:via, :swarm, Test.Scheduler.TaskRegistry}, task_supervisor: Test.Scheduler.Task.Supervisor}]}, :temporary, 5000, :worker, [Quantum.Executor]}}
 
20:30:17.647 [info]  GenStage consumer Test.Scheduler.ExecutorSupervisor is stopping after receiving cancel from producer #PID<0.190.0> with reason: :noproc

 
20:30:17.648 [error] GenServer Test.Scheduler.ExecutorSupervisor terminating
** (stop) no process: the process is not alive or there's no process currently associated with the given name, possibly because its application isn't started
Last message: {:DOWN, #Reference<0.3155748787.2694316036.156586>, :process, #PID<0.190.0>, :noproc}
State: %ConsumerSupervisor{args: %{cluster_task_supervisor_registry: Test.Scheduler.ClusterTaskSupervisorRegistry, debug_logging: true, execution_broadcaster: {:via, :swarm, Test.Scheduler.ExecutionBroadcaster}, task_registry: {:via, :swarm, Test.Scheduler.TaskRegistry}, task_supervisor: Test.Scheduler.Task.Supervisor}, children: %{}, max_restarts: 3, max_seconds: 5, mod: Quantum.ExecutorSupervisor, name: Test.Scheduler.ExecutorSupervisor, producers: %{}, restarting: 0, restarts: [], strategy: :one_for_one, template: {Quantum.Executor, {Quantum.Executor, :start_link, [%{cluster_task_supervisor_registry: Test.Scheduler.ClusterTaskSupervisorRegistry, debug_logging: true, task_registry: {:via, :swarm, Test.Scheduler.TaskRegistry}, task_supervisor: Test.Scheduler.Task.Supervisor}]}, :temporary, 5000, :worker, [Quantum.Executor]}}
 
20:30:17.648 [debug] [swarm on b@127.0.0.1] [tracker:handle_call] registering #PID<0.218.0> as :"Elixir.Test.Scheduler.ClusterTaskSupervisorRegistry.Group.#Reference<0.3155748787.2694316033.156137>", with metadata %{}

20:30:17.648 [info]  GenStage consumer Test.Scheduler.ExecutorSupervisor is stopping after receiving cancel from producer #PID<0.190.0> with reason: :noproc

 
20:30:17.649 [error] GenServer Test.Scheduler.ExecutorSupervisor terminating
** (stop) no process: the process is not alive or there's no process currently associated with the given name, possibly because its application isn't started
Last message: {:DOWN, #Reference<0.3155748787.2694316036.156588>, :process, #PID<0.190.0>, :noproc}
State: %ConsumerSupervisor{args: %{cluster_task_supervisor_registry: Test.Scheduler.ClusterTaskSupervisorRegistry, debug_logging: true, execution_broadcaster: {:via, :swarm, Test.Scheduler.ExecutionBroadcaster}, task_registry: {:via, :swarm, Test.Scheduler.TaskRegistry}, task_supervisor: Test.Scheduler.Task.Supervisor}, children: %{}, max_restarts: 3, max_seconds: 5, mod: Quantum.ExecutorSupervisor, name: Test.Scheduler.ExecutorSupervisor, producers: %{}, restarting: 0, restarts: [], strategy: :one_for_one, template: {Quantum.Executor, {Quantum.Executor, :start_link, [%{cluster_task_supervisor_registry: Test.Scheduler.ClusterTaskSupervisorRegistry, debug_logging: true, task_registry: {:via, :swarm, Test.Scheduler.TaskRegistry}, task_supervisor: Test.Scheduler.Task.Supervisor}]}, :temporary, 5000, :worker, [Quantum.Executor]}}
 
20:30:17.649 [debug] [swarm on b@127.0.0.1] [tracker:handle_call] add_meta {Test.Scheduler.ClusterTaskSupervisorRegistry.Group, true} to #PID<0.218.0>
 
20:30:17.650 [debug] [swarm on b@127.0.0.1] [tracker:handle_monitor] :"Elixir.Test.Scheduler.ClusterTaskSupervisorRegistry.Group.#Reference<0.3155748787.2694316033.156137>" is down: :shutdown
 
20:30:17.651 [debug] [swarm on b@127.0.0.1] [tracker:handle_replica_event] replica event: untrack #PID<0.218.0>
 
20:30:17.652 [info]  Application test exited: shutdown
{"Kernel pid terminated",application_controller,"{application_terminated,test,shutdown}"}
Kernel pid terminated (application_controller) ({application_terminated,test,shutdown})

Crash dump is being written to: erl_crash.dump...done
```

FYI.

**Comment by maennchen** [comment]
I did a really dirty attempted fix. Could both @redink  and @zacksiri try the newest commit on the PR?

**Comment by zacksiri** [comment]
@maennchen just about to hit the hay, I will definitely give this a whirl tomorrow.

**Comment by maennchen** [comment]
@zacksiri Ok, then I'll have to figure out how to do this nicely instead of that hack.

**Comment by zacksiri** [comment]
@maennchen thx for your hard work! will try this out soon.

**Comment by zacksiri** [comment]
@maennchen I tried this out on my staging server, everything looks good!! I think you can merge it.

**Comment by maennchen** [comment]
Released with `v2.3.2`.

**Comment by redink** [comment]
@maennchen Thanks very much, I will test this release in our env. Thanks again.
